src/Parser/Core/Modules/Items/GemChecker.js

```diff
@@ -0,0 +1,176 @@
+import React from 'react';
+
+import Analyzer from 'Parser/Core/Analyzer';
+import SUGGESTION_IMPORTANCE from 'Parser/Core/ISSUE_IMPORTANCE';
+import ITEMS from 'common/ITEMS';
+import ItemLink from 'common/ItemLink';
+
+const GEMS = [
+  ITEMS.DEADLY_SOLSTONE.id,
+  ITEMS.QUICK_GOLDEN_BERYL.id,
+  ITEMS.MASTERFUL_KUBILINE.id,
+  ITEMS.VERSATILE_KYANITE.id,
+];
+
+const BEST_GEMS = [
+  ITEMS.DEADLY_AMBERBLAZE.id,
+  ITEMS.QUICK_OWLSEYE.id,
+  ITEMS.MASTERFUL_TIDAL_AMETHYST.id,
+  ITEMS.VERSATILE_ROYAL_QUARTZ.id,
+];
+
+const UNIQUE_GEMS = [
+  ITEMS.KRAKEN_EYE_OF_AGILITY.id,
+  ITEMS.KRAKEN_EYE_OF_INTELLECT.id,
+  ITEMS.KRAKEN_EYE_OF_STRENGTH.id,
+];
+
+class GemChecker extends Analyzer {
+
+  static GEMMABLE_SLOTS = {
+    0: 'Head',
+    2: 'Shoulders',
+    4: 'Chest',
+    5: 'Waist',
+    6: 'Legs',
+    7: 'Feet',
+    8: 'Wrists',
+    9: 'Hands',
+    10: 'Ring',
+    11: 'Ring',
+    12: 'Trinket',
+    13: 'Trinket',
+    14: 'Back',
+    15: 'MainHand',
+    16: 'OffHand',
+  };
+
+  gemmableGear() {
+    return Object.keys(this.constructor.GEMMABLE_SLOTS).reduce((obj, slot) => {
+      obj[slot] = this.selectedCombatant._getGearItemBySlotId(slot);
+      return obj;
+    }, {});
+  }
+
+  isAtLeastGemmed(id) {
+    return GEMS.includes(id) || this.isMaxGemmed(id);
+  }
+
+  isMaxGemmed(id) {
+    return BEST_GEMS.includes(id) || this.isUniqueGemmed(id);
+  }
+
+  isUniqueGemmed(id) {
+    return UNIQUE_GEMS.includes(id);
+  }
+
+  hasEveryGemSlotFilled(item) {
+    let everySlotIsGemmed = true;
+    item.gems.forEach(({ id }, index) => {
+      everySlotIsGemmed &= this.isAtLeastGemmed(id);
+    });
+    return everySlotIsGemmed;
+  }
+
+  hasEveryGemSlotMaxFilled(item) {
+    let everySlotIsMaxGemmed = true;
+    item.gems.forEach(({ id }, index) => {
+      everySlotIsMaxGemmed &= this.isMaxGemmed(id);
+    });
+    return everySlotIsMaxGemmed;
+  }
+
+  get numSockets() {
+    const gear = this.gemmableGear;
+    let result = 0;
+    Object.keys(gear).forEach((slot) => {
+      result += gear[slot].gems.length;
+    });
+    return result;
+  }
+
+  get numSocketsMissingGem() {
+    const gear = this.gemmableGear;
+    let result = 0;
+    Object.values(gear).forEach((item) => {
```

Comment:
things get a bit ugly here, but this should be a reducer within a reducer instead of a forEach within a forEach modifying a `let` variables two scopes above it